// formik-chakra-ui/dist/formik-chakra-ui.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var react = require('@chakra-ui/react');
var React = require('react');
var React__default = _interopDefault(React);
var formik = require('formik');
var react$1 = require('@emotion/react');

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function _taggedTemplateLiteralLoose(strings, raw) {
  if (!raw) {
    raw = strings.slice(0);
  }

  strings.raw = raw;
  return strings;
}

var FormControl = function FormControl(props) {
  var children = props.children,
      name = props.name,
      label = props.label,
      labelProps = props.labelProps,
      helperText = props.helperText,
      helperTextProps = props.helperTextProps,
      errorMessageProps = props.errorMessageProps,
      rest = _objectWithoutPropertiesLoose(props, ["children", "name", "label", "labelProps", "helperText", "helperTextProps", "errorMessageProps"]);

  var _useField = formik.useField(name),
      _useField$ = _useField[1],
      error = _useField$.error,
      touched = _useField$.touched;

  return React__default.createElement(react.FormControl, Object.assign({
    isInvalid: !!error && touched
  }, rest), label && React__default.createElement(react.FormLabel, Object.assign({
    htmlFor: name
  }, labelProps), label), children, error && React__default.createElement(react.FormErrorMessage, Object.assign({}, errorMessageProps), error), helperText && React__default.createElement(react.FormHelperText, Object.assign({}, helperTextProps), helperText));
};

var CheckboxContainer = function CheckboxContainer(props) {
  var name = props.name,
      label = props.label,
      children = props.children,
      stackProps = props.stackProps,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "children", "stackProps"]);

  return React__default.createElement(FormControl, Object.assign({
    name: name,
    label: label
  }, rest), React__default.createElement(react.Stack, Object.assign({
    pl: 6,
    mt: 1,
    spacing: 1
  }, stackProps), children));
};

var CheckboxControl = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var name = props.name,
      label = props.label,
      children = props.children,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "children"]);

  var _useField = formik.useField(name),
      field = _useField[0],
      _useField$ = _useField[1],
      error = _useField$.error,
      touched = _useField$.touched;

  var isChecked;

  if (field.value instanceof Array) {
    var _field$value$includes;

    isChecked = (_field$value$includes = field.value.includes(props.value)) != null ? _field$value$includes : false;
  }

  return React__default.createElement(react.Checkbox, Object.assign({}, field, {
    isInvalid: !!error && touched,
    isChecked: isChecked,
    ref: ref
  }, rest), label, children);
});

var CheckboxSingleControl = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var name = props.name,
      label = props.label,
      children = props.children,
      checkBoxProps = props.checkBoxProps,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "children", "checkBoxProps"]);

  var _useField = formik.useField(name),
      field = _useField[0],
      _useField$ = _useField[1],
      error = _useField$.error,
      touched = _useField$.touched;

  var isChecked = field.value;
  return React__default.createElement(FormControl, Object.assign({
    name: name
  }, rest), React__default.createElement(react.Checkbox, Object.assign({}, field, {
    id: name,
    isInvalid: !!error && touched,
    isChecked: isChecked,
    ref: ref
  }, checkBoxProps), label, children));
});

var InputControl = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var name = props.name,
      label = props.label,
      inputProps = props.inputProps,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "inputProps"]);

  var _useField = formik.useField(name),
      field = _useField[0];

  return React__default.createElement(FormControl, Object.assign({
    name: name,
    label: label
  }, rest), React__default.createElement(react.Input, Object.assign({}, field, {
    id: name
  }, inputProps, {
    ref: ref
  })));
});

var NumberInputControl = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var name = props.name,
      label = props.label,
      _props$showStepper = props.showStepper,
      showStepper = _props$showStepper === void 0 ? true : _props$showStepper,
      children = props.children,
      numberInputProps = props.numberInputProps,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "showStepper", "children", "numberInputProps"]);

  var _useField = formik.useField(name),
      field = _useField[0],
      _useField$ = _useField[1],
      error = _useField$.error,
      touched = _useField$.touched;

  var _useFormikContext = formik.useFormikContext(),
      setFieldValue = _useFormikContext.setFieldValue;

  var $setFieldValue = function $setFieldValue(name) {
    return function (value) {
      return setFieldValue(name, value);
    };
  };

  return React__default.createElement(FormControl, Object.assign({
    name: name,
    label: label
  }, rest), React__default.createElement(react.NumberInput, Object.assign({}, field, {
    id: name,
    onChange: $setFieldValue(name),
    isInvalid: !!error && touched
  }, numberInputProps), React__default.createElement(react.NumberInputField, {
    name: name,
    ref: ref
  }), showStepper && React__default.createElement(react.NumberInputStepper, null, React__default.createElement(react.NumberIncrementStepper, null), React__default.createElement(react.NumberDecrementStepper, null)), children));
});

var calculateProgress = function calculateProgress(numFields, numErrors) {
  var validFields = numFields - numErrors;
  return validFields / numFields * 100;
};

var PercentComplete = function PercentComplete(props) {
  var _props$progressFn = props.progressFn,
      progressFn = _props$progressFn === void 0 ? calculateProgress : _props$progressFn,
      progressProps = props.progressProps,
      rest = _objectWithoutPropertiesLoose(props, ["progressFn", "progressProps"]);

  var _useFormikContext = formik.useFormikContext(),
      errors = _useFormikContext.errors,
      values = _useFormikContext.values,
      validateForm = _useFormikContext.validateForm,
      dirty = _useFormikContext.dirty;

  var numFields = Object.keys(values).length;
  var numErrors = Object.keys(errors).length;
  React.useEffect(function () {
    validateForm();
  }, [dirty]);
  return React__default.createElement(react.Box, Object.assign({
    marginY: 5
  }, rest), React__default.createElement(react.Progress, Object.assign({
    hasStripe: true,
    isAnimated: true,
    value: progressFn(numFields, numErrors)
  }, progressProps)));
};

var RadioGroupControl = function RadioGroupControl(props) {
  var name = props.name,
      label = props.label,
      radioGroupProps = props.radioGroupProps,
      stackProps = props.stackProps,
      children = props.children,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "radioGroupProps", "stackProps", "children"]);

  var _useField = formik.useField(name),
      field = _useField[0];

  var _useFormikContext = formik.useFormikContext(),
      setFieldValue = _useFormikContext.setFieldValue;

  var handleChange = function handleChange(value) {
    setFieldValue(name, value);
  };

  return React__default.createElement(FormControl, Object.assign({
    name: name,
    label: label
  }, rest), React__default.createElement(react.RadioGroup, Object.assign({}, field, {
    onChange: handleChange
  }, radioGroupProps), React__default.createElement(react.Stack, Object.assign({
    direction: "row"
  }, stackProps), children)));
};

var ResetButton = function ResetButton(props) {
  var children = props.children,
      rest = _objectWithoutPropertiesLoose(props, ["children"]);

  var _useFormikContext = formik.useFormikContext(),
      isSubmitting = _useFormikContext.isSubmitting,
      dirty = _useFormikContext.dirty,
      resetForm = _useFormikContext.resetForm;

  return React__default.createElement(react.Button, Object.assign({
    colorScheme: "teal",
    variant: "outline",
    onClick: function onClick() {
      return resetForm();
    },
    isDisabled: isSubmitting || !dirty
  }, rest), children);
};

var SelectControl = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var name = props.name,
      label = props.label,
      selectProps = props.selectProps,
      children = props.children,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "selectProps", "children"]);

  var _useField = formik.useField(name),
      field = _useField[0];

  return React__default.createElement(FormControl, Object.assign({
    name: name,
    label: label
  }, rest), React__default.createElement(react.Select, Object.assign({}, field, {
    id: name,
    ref: ref
  }, selectProps), children));
});

var SubmitButton = function SubmitButton(props) {
  var children = props.children,
      rest = _objectWithoutPropertiesLoose(props, ["children"]);

  var _useFormikContext = formik.useFormikContext(),
      isSubmitting = _useFormikContext.isSubmitting;

  return React__default.createElement(react.Button, Object.assign({
    type: "submit",
    isLoading: isSubmitting,
    colorScheme: "teal"
  }, rest), children);
};

var _templateObject;
var SwitchControl = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var name = props.name,
      label = props.label,
      switchProps = props.switchProps,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "switchProps"]);

  var _useField = formik.useField(name),
      field = _useField[0],
      _useField$ = _useField[1],
      error = _useField$.error,
      touched = _useField$.touched;

  return React__default.createElement(react.Box, {
    css: react$1.css(_templateObject || (_templateObject = _taggedTemplateLiteralLoose(["\n          .chakra-form__label {\n            margin-bottom: 0;\n          }\n          .chakra-switch {\n            display: flex;\n            align-items: center;\n            margin-right: 0.75rem;\n          }\n          .chakra-form__error-message {\n            margin-top: 0;\n          }\n        "])))
  }, React__default.createElement(FormControl, Object.assign({
    name: name,
    label: label,
    as: react.Flex,
    alignItems: "center"
  }, rest), React__default.createElement(react.Switch, Object.assign({}, field, {
    id: name,
    isInvalid: !!error && touched,
    isChecked: field.value,
    ref: ref
  }, switchProps))));
});

var TextareaControl = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var name = props.name,
      label = props.label,
      textareaProps = props.textareaProps,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "textareaProps"]);

  var _useField = formik.useField(name),
      field = _useField[0];

  return React__default.createElement(FormControl, Object.assign({
    name: name,
    label: label
  }, rest), React__default.createElement(react.Textarea, Object.assign({}, field, {
    id: name,
    ref: ref
  }, textareaProps)));
});

var SliderControl = function SliderControl(props) {
  var name = props.name,
      label = props.label,
      sliderProps = props.sliderProps,
      sliderTrackProps = props.sliderTrackProps,
      sliderThumbProps = props.sliderThumbProps,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "sliderProps", "sliderTrackProps", "sliderThumbProps"]);

  var _useField = formik.useField(name),
      field = _useField[0],
      setValue = _useField[2].setValue;

  function handleChange(value) {
    setValue(value);
  } // Does not behave like expected, so we manually handle it.


  function handleBlur(e) {
    e.target.name = name;
    field.onBlur(e);
  }

  return React__default.createElement(FormControl, Object.assign({
    name: name,
    label: label
  }, rest), React__default.createElement(react.Slider, Object.assign({}, field, {
    id: name,
    onChange: handleChange,
    onBlur: handleBlur
  }, sliderProps), React__default.createElement(react.SliderTrack, Object.assign({}, sliderTrackProps), React__default.createElement(react.SliderFilledTrack, null)), React__default.createElement(react.SliderThumb, Object.assign({}, sliderThumbProps))));
};

var PinInputControl = function PinInputControl(props) {
  var name = props.name,
      label = props.label,
      pinAmount = props.pinAmount,
      stackProps = props.stackProps,
      pinInputProps = props.pinInputProps,
      rest = _objectWithoutPropertiesLoose(props, ["name", "label", "pinAmount", "stackProps", "pinInputProps"]);

  var _useField = formik.useField(name),
      field = _useField[0],
      setValue = _useField[2].setValue;

  var renderedPinInputFields = Array(pinAmount).fill(null).map(function (_noop, i) {
    return React__default.createElement(react.PinInputField, {
      key: i
    });
  });

  function handleChange(value) {
    setValue(value);
  }

  return React__default.createElement(FormControl, Object.assign({
    name: name,
    label: label
  }, rest), React__default.createElement(react.HStack, Object.assign({}, stackProps), React__default.createElement(react.PinInput, Object.assign({}, field, {
    onChange: handleChange
  }, pinInputProps), renderedPinInputFields)));
};

exports.CheckboxContainer = CheckboxContainer;
exports.CheckboxControl = CheckboxControl;
exports.CheckboxSingleControl = CheckboxSingleControl;
exports.FormControl = FormControl;
exports.InputControl = InputControl;
exports.NumberInputControl = NumberInputControl;
exports.PercentComplete = PercentComplete;
exports.PinInputControl = PinInputControl;
exports.RadioGroupControl = RadioGroupControl;
exports.ResetButton = ResetButton;
exports.SelectControl = SelectControl;
exports.SliderControl = SliderControl;
exports.SubmitButton = SubmitButton;
exports.SwitchControl = SwitchControl;
exports.TextareaControl = TextareaControl;
